queenbee/io/alias.py

```diff
@@ -0,0 +1,568 @@
+from jsonschema import validate as json_schema_validator
+
+from .common import ItemType, GenericInput, FromOutput, find_dup_items
+from .artifact_source import HTTP, S3, ProjectFolder
+from ..io.reference import FileReference, FolderReference, TaskReference
+from ..base.basemodel import BaseModel
+
+
+class IOAliasHandler(BaseModel):
+    """Input and output alias handler object."""
+
+    type: constr(regex='^IOAliasHandler$') = 'IOAliasHandler'
+
+    language: str = Field(
+        ...,
+        description='Declare the language (e.g. python, csharp, etc.). This '
+        'option allows the recipe to be flexible on handling different programming '
+        'languages.'
+    )
+
+    module: str = Field(
+        ...,
+        description='Target module or namespace to load the alias function.',
+        example='honeybee_rhino.handlers'
+    )
+
+    function: str = Field(
+        ...,
+        description='Name of the function. The input value will be passed to this '
+        'function as the first argument.'
+    )
+
+
+class DAGGenericInputAlias(GenericInput):
+    """Base class for DAG Alias inputs.
+
+    This class adds a handler to input to handle the process of loading the input
+    from different graphical interfaces.
+    """
+    type: constr(regex='^DAGGenericInputAlias$') = 'DAGGenericInputAlias'
+
+    platform: List[str] = Field(
+        ...,
+        description='Name of the client platform (e.g. Grasshopper, Revit, etc). The '
+        'value can be any strings as long as it has been agreed between client-side '
+        'developer and author of the recipe.'
+    )
+
+    handler: List[IOAliasHandler] = Field(
+        ...,
+        description='List of process actions to process the input or output value.'
+    )
+
+    @validator('platform', always=True)
+    def create_empty_platform_list(cls, v):
+        return [] if v is None else v
+
+    @validator('handler', always=True)
+    def check_duplicate_platform_name(cls, v, values):
+        v = [] if v is None else v
+        languages = [h.language for h in v]
+        dup_lang = find_dup_items(languages)
+        if dup_lang:
+            raise ValueError(
+                f'Duplicate use of language(s) found in alias handlers for '
+                f'{values["platform"]}: {dup_lang}. Each language can only be used once '
+                'in each platform.'
+            )
+        return v
+
+
+class DAGStringInputAlias(DAGGenericInputAlias):
+    """An Alias String input.
+
+    You can add additional validation by defining a JSONSchema specification.
+
+    See http://json-schema.org/understanding-json-schema/reference/string.html#string for
+    more information.
+
+    .. code-block:: python
+
+        "schema": {
+            "type": "string",
+            "maxLength": 50,
+            "pattern": "(?i)(^.*\\.epw$)"
+        }
+
+    """
+
+    type: constr(regex='^DAGStringInputAlias$') = 'DAGStringInputAlias'
+
+    default: str = Field(
```

Comment:
Why do we need the default here in alias object, can't we just use the default from upper non-alias object?

Reply:
This will be useful is the alias type is different from the original input. In those cases we need to be able to provide a default value from the same type of the alias input.